player/tests/test.rs

```diff
@@ -0,0 +1,169 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/*! Tester for WebGPU
+ *  It enumerates the available backends on the system,
+ *  and run the tests through them.
+ *
+ *  Test requirements:
+ *    - all IDs have the backend `Empty`
+ *    - all expected buffers have `MAP_READ` usage
+ *    - last action is `Submit`
+!*/
+
+use player::{gfx_select, GlobalPlay, IdentityPassThroughFactory};
+use std::{
+    fs::{read_to_string, File},
+    path::{Path, PathBuf},
+    ptr, slice,
+};
+
+#[derive(serde::Deserialize)]
+struct RawId {
+    index: u32,
+    epoch: u32,
+}
+
+#[derive(serde::Deserialize)]
+struct Expectation {
+    name: String,
+    buffer: RawId,
+    offset: wgt::BufferAddress,
+    data: Vec<u8>,
+}
+
+#[derive(serde::Deserialize)]
+struct Test {
+    path: String,
+    features: wgt::Features,
+    expectations: Vec<Expectation>,
+}
+
+#[derive(serde::Deserialize)]
+struct Corpus {
+    backends: wgt::BackendBit,
+    tests: Vec<Test>,
+}
+
+const BACKENDS: &[wgt::Backend] = &[
+    wgt::Backend::Vulkan,
+    wgt::Backend::Metal,
+    wgt::Backend::Dx12,
+    wgt::Backend::Dx11,
+    wgt::Backend::Gl,
+];
+
+extern "C" fn map_callback(status: wgc::resource::BufferMapAsyncStatus, _user_data: *mut u8) {
+    match status {
+        wgc::resource::BufferMapAsyncStatus::Success => (),
+        _ => panic!("Unable to map"),
+    }
+}
+
+fn run_test(
+    test: Test,
+    dir: &Path,
+    global: &wgc::hub::Global<IdentityPassThroughFactory>,
+    adapter: wgc::id::AdapterId,
+) {
+    println!("\t\tTest '{:?}'", test.path);
+    let device = gfx_select!(adapter => global.adapter_request_device(
+        adapter,
+        &wgt::DeviceDescriptor {
+            features: test.features | wgt::Features::MAPPABLE_PRIMARY_BUFFERS,
```

Comment:
Is this possible on discrete GPUs?

Reply:
It's always allowed (mainly as I want to always have it on for _reasons_) but wgpu log::warn if you turn it on with a discrete gpu.

Reply:
What it means is that the buffer will live on the system memory side but still be accessible by the GPU. I think it's fine here, especially for the first iteration.